### PR TITLE
added einsum op and allow more broadcasting dims

### LINKED ISSUE #175: einsum-like function?
Numpy has einsum which is very useful for formulating and computing tensor operations efficiently. Any plans to include an equivalent in tensorflow?

### PR DESCRIPTION
**Pull request by nikhilmishra000**
This was meant to solve: https://github.com/tensorflow/tensorflow/issues/175
And I think it addresses this too: https://github.com/tensorflow/tensorflow/issues/1519

Fixes #175, closing #1519.

### CODE REVIEW

**Comment by andrewharp** [comment]
@tensorflow-jenkins Test this please.

**Comment by nikhilmishra000** [comment]
I'm not too sure why `//tensorflow/contrib/learn:graph_io_test` is failing on python 3 only. Is there a log file from that specific test that I could take a look at? I can't seem to reproduce the failure locally...

**Comment by andrewharp** [comment]
The //tensorflow/contrib/learn:graph_io_test failure is most likely unrelated. It's already failing on the nightly builds for Mac python3.

File: tensorflow/python/ops/special_math_ops.py [hunk 1]
```diff
@@ -86,3 +88,75 @@ def empty_lbeta():
         return empty_lbeta()
     else:
       return control_flow_ops.cond(is_empty, empty_lbeta, nonempty_lbeta)
+
+
+def einsum(axes, *inputs):
+  r"""
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops.py]: You don't have escapes in this string, so remove the r.

File: tensorflow/python/ops/special_math_ops_test.py [hunk 2]
```diff
@@ -111,5 +111,116 @@ class LBetaTestGpu(LBetaTest):
...
+      output_value = output_tensor.eval()
+
+    correct_value = np.einsum(axes, *input_vals)
+
+    err = np.abs(correct_value - output_value).max()
+    print(axes, err)
+    assert err < 1e-8
+
+
+class EinsumTestGpu(EinsumTest):
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops_test.py]: You should remove this class, and always enable processing on the GPU. We run the tests on CPU only and GPU machines. The other tests using the same pattern are vestigial.

**Review by martinwicke — CHANGES_REQUESTED** [review]
LGTM. Small changes. Sorry about the GPU confusion, we're in the process of changing how we do GPU tests, hence those leftover things. @gunan FYI.

**Comment by nikhilmishra000** [comment]
@martinwicke done

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Thanks! Jenkins, test this please.

**Comment by nikhilmishra000** [comment]
only failure is `//tensorflow/contrib/learn:graph_io_test`, which we said was unrelated

**Comment by martinwicke** [comment]
Hooray.

**Comment by pannous** [comment]
Fantastic! 
Any reasons against an alias 'contraction' ?

**Comment by nikhilmishra000** [comment]
nope, I had named it `einsum` because of similarity to `numpy.einsum`

**Pull request merged**